### DISCUSSION TITLE
variable interpolation

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#316] [Ideas & RFCs]
This has been converted to an RFC: [RFC-3: Variable Interpolation In Strings](https://fastn.com/rfc/variable-interpolation/).

Feel free to post your comments here.

We want this to work:

```ftd
-- import: fpm

-- string name: Jack

-- ftd.text: hello, {$name}
src: //{$fpm.package.name}
```

**Comment by amitu** [comment]
We should use Java like safe interpolation: https://openjdk.org/jeps/430

**Comment by Arpita-Jaiswal** [comment]
```ftd
-- string a: hello ${ name + last-name + "good" }
-- ftd.text: hello ${ 
   name + last-name + bar(a = name) + ftd.to-string(a = age) +
   foo(a = {bar(a= name)}, b = "good", c = {foo(a="hello") + " the end!"}) 
}

-- string name: John
-- string last-name: Doe
-- integer age: 10
```

```html
<div>
"Arpita jaiswal"
<div>Arpita is good</div>
</div>
```

```ftd
-- component foo:
caption a:
optional string b:
optional string c:

<implementation omitted>

-- string bar():
caption a:

<implementation omitted>
```

**Reply by amitu** [reply]
@Arpita-Jaiswal: `ftd.to-string(a = age)` means `ftd.to-string` is a generic, can take data of any type. Other alternative is Elm like `String.fromInt()` etc, no need for generic. I tend to like the second approach.

**Reply by amitu** [reply]
Also I feel in our function, let's allow people to declare one with type `caption`, which can have the name omitted. It will allow us to write `ftd.integer-to-string(10)` without needing the name: ``ftd.integer-to-string(a = 10)`. 

I am not too fond of the variables named `a` btw 😛.
